### DISCUSSION TITLE
How to combine mulitipanelplot objects into 1 page and label them A, B, C ...?

### DISCUSSION THREAD

**Discussion by jarbet** [uclahs-cds/public-R-BoutrosLab-plotting-general#135] [Q&A] [answered]
I have previously used `cowplot::plot_grid` and `gridExtra::grid.arrange` to combine multiple BPG plots into a single page, and give the plots labels, e.g. **A, B, C**..

However, this does not work for plots made from `create.multipanelplot`.

I am currently trying this [option](https://github.com/uclahs-cds/project-disease-ProstateTumor-PRAD-000095-ColibactinDHTDNARNA/blob/main/scripts/manuscript/figures/Figure1_Overview.r) from @raagagrawal (write plots to disk, read them in using cowplot, `plot_grid` to combine, then `save_plot`), which I implemented [here](https://github.com/uclahs-cds/project-disease-ProstateTumor-PRAD-000101-MethySubtypes/blob/a65e156a69a8c03eaa94f5e0d170c7532b2e0b1f/analysis/driver_mutations/driverMutations_vs_methy.Rmd#L558-L582).  

However, I am curious if anyone has other solutions for combining multipanelplots into 1 page and labeling them?

**Comment by pboutros** [comment]
Yup a couple of suggestions.  In `write.plot()` there is the capability to pass in additional plots and specifically locate them via `additional.trellis.objects` and `additional.trellis.locations`. IIRC that was particularly designed for insets, but is likely useful for lots of other things.

But if that doesn't work, I believe the general solution is to do `plot(trells.object, newpage = TRUE);` for the first one and then for subsequent ones `plot(trellis.objection, position = c(coordinates), newpage = FALSE);`  This should be how `write.plot()` actually implements the above.

I believe there's an example of the first in `write.plot()`, and then suggest opening a PR to add an example of the second to `write.plot()` as well

**Reply by jarbet** [reply]
Thanks @pboutros.  I got your approach to work for overlaying 2 plots into one plot, but what if you want to arrange plots side-by-side and label them A,B,C... something like this:

![image](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/assets/103967480/72230604-b3e8-411f-910e-fc67603a1192)

Do you know if your approach supports that?

At a minimum, I was able to get Raag's solution to work, but still curious about other options.

**Comment by pboutros** [comment]
It definitely does -- there should be examples on our internal github repo to find, and that's the standard mechanism we'll want to be using.

**Comment by stefaneng** [comment]
> However, this does not work for plots made from create.multipanelplot.

@jarbet 
You can actually put multipanelplots together with `grid.arrange` but the class needs to change. 
`create.multipanelplot` removes the other grob associated class (expect to print: https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/blob/main/R/create.multipanelplot.R#L6C17-L6C53). Also, once you change the class to a frame then you can annotate the grob with `placeGrob` and add text via a `textGrob`

For example:
```r
library(BoutrosLab.plotting.general);

set.seed(13);
sample.data <- data.frame(x = rnorm(100), y = rnorm(100))
plot1 <- create.scatterplot(
    y ~ x,
    data = sample.data
    );

plot2 <- create.histogram(sample.data$y);

plot1.mpp <- create.multipanelplot(list(plot1), layout.height = 1, layout.width = 1);
plot2.mpp <- create.multipanelplot(list(plot2), layout.height = 1, layout.width = 1);
class(plot1.mpp) <- c('multipanel', 'frame', 'gTree', 'grob', 'gDesc');
class(plot2.mpp) <- c('multipanel', 'frame', 'gTree', 'grob', 'gDesc');

plot1.anno.mpp <- placeGrob(plot1.mpp, textGrob('A', x = 0.1, y = 0.9, gp = gpar(fontface = 'bold', cex = 2)));
plot2.anno.mpp <- placeGrob(plot2.mpp, textGrob('B', x = 0.1, y = 0.9, gp = gpar(fontface = 'bold', cex = 2)));

combined.fig <- gridExtra::grid.arrange(plot1.anno.mpp, plot2.anno.mpp, ncol = 2);

write.plot(
    combined.fig,
    width = 12,
    height = 6,
    filename = 'test.png',
    resolution = 100
    );
```

![image](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/assets/1639487/ec27ad4d-8549-4191-802c-38c54df57bfa)


I wasn't able to do it via `write.plot` either. The `position` argument from `print.trellis` seems to only place the plot within the current one. This isn't to say it isn't possible, I just couldn't figure out how to do it in my quick search. Github search didn't turn up any examples of this either.

**Comment by jarbet** [comment, marked as answer]
Although I have used `cowplot::plot_grid` in the past to combine figures and label them (e.g. a,b,c,d), I recently ran into a scenario where this did not work and I needed to use `create.multipanelplot` to combine the figures instead.

I can confirm that @stefaneng's above approach works very well!   I've simplified it a bit below: 

### Labeling sub-plots when using `create.multipanelplot`:

1. Make your bpg plots
2. Make sure the class of each plot is correct, *e.g.*:

```
class(plot1.mpp) <- c('frame', 'gTree', 'grob', 'gDesc');
```

Note if the plots you are combining are each *multipanelplots*, then you need to add class `multipanelplot` above

3. Add the labels to each plot, *e.g.*:

```
plot1.anno.mpp <- placeGrob(plot1.mpp, textGrob('A', x = 0.1, y = 0.9, gp = gpar(fontface = 'bold', cex = 2)));
```

4. Combine the new 'annotated' plots using `create.multipanelplot`

---

A major benefit of this approach is you can add an Overall legend using `create.multipanelplot`, whereas you cannot do this with `cowplot::plot_grid`

`gridExtra` is not needed, the entire above process can be done in BPG
